Lab này sẽ giới thiệu ngắn gọn về Python. Ngay cả khi bạn đã sử dụng Python trước đó, điều này sẽ giúp bạn làm quen với các hàm mà chúng ta sẽ cần.

**Hướng dẫn:**
- Chúng ta sẽ sử dụng Python 3.
- Tránh sử dụng vòng lặp for và vòng lặp while, trừ khi được yêu cầu rõ ràng làm như vậy.
- Sau khi mã hóa hàm của bạn, hãy chạy cell ngay bên dưới để kiểm tra xem kết quả của bạn có đúng không.

**Sau lab này, chúng ta sẽ:**
- Có thể sử dụng iPython Notebook
- Có thể dùng các hàm numpy và các phép toán ma trận/vectơ numpy
- Hiểu được khái niệm "broadcasting"
- Có thể vectơ hóa code

Hãy bắt đầu nào!

## Về iPython Notebook ##

iPython Notebooks là môi trường mã hóa tương tác được nhúng trong một trang web. Bạn sẽ sử dụng iPython notebook trong lab này. Bạn chỉ cần viết code giữa chú thích ### BẮT ĐẦU CODE Ở ĐÂY ### và ### KẾT THÚC CODE Ở ĐÂY ###. Sau khi viết code, bạn có thể chạy cell bằng cách nhấn "SHIFT"+"ENTER" hoặc nhấp vào "Run Cell" (được biểu thị bằng biểu tượng) ở thanh trên của notebook.

Chúng ta thường chỉ định "(≈ X dòng code)" trong các chú thích để cho bạn biết bạn cần viết bao nhiêu code. Đó chỉ là một ước tính sơ bộ, vì vậy đừng cảm thấy tồi tệ nếu code của bạn dài hoặc ngắn hơn.

**Task 1**: Đặt test thành `"Hello World"` trong cell bên dưới để in "Hello World" và chạy hai cell bên dưới.

In [1]:
### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 1 dòng code)
test = "Hello World"
### KẾT THÚC CODE Ở ĐÂY ###

In [2]:
print ("test: " + test)

test: Hello World


**Kỳ vọng đầu ra**:
test: Hello World

<font color='blue'>
**Hãy nhớ**:
- Chạy cell bằng cách nhấn SHIFT+ENTER (hoặc "Run cell")
- Chỉ viết code trong các vùng chỉ định dùng Python 3
- Không sửa đổi code bên ngoài vùng chỉ định

## 1 - Tạo các hàm cơ bản với numpy ##

Numpy là thư viện chính cho tính toán khoa học trong Python. Nó được duy trì bởi một cộng đồng lớn (www.numpy.org). Trong lab này, bạn sẽ học một số hàm numpy chính như np.exp, np.log và np.reshape. Bạn sẽ cần biết cách sử dụng các hàm này cho các bài tập trong tương lai.

### 1.1 - sigmoid function, np.exp() ###

Trước khi sử dụng np.exp(), bạn sẽ sử dụng math.exp() để triển khai hàm sigmoid. Sau đó, bạn hiểu tại sao np.exp() lại thích hợp hơn math.exp().

**Task 2**: Xây dựng một hàm trả về giá trị sigmoid của số thực x. Sử dụng math.exp(x) cho hàm mũ.

**Nhắc nhở**:
$sigmoid (x) = \frac{1}{1+e^ {- x}}$ đôi khi còn được gọi là hàm logistic. Đây là một hàm phi tuyến tính không chỉ được sử dụng trong Học máy (Hồi quy logistic) mà còn trong Học sâu (Deep Learning).

<img src="images/Sigmoid.png" style="width:500px;height:228px;">

Để tham chiếu đến một hàm thuộc một thư viện cụ thể, bạn có thể gọi nó bằng cách sử dụng package_name.functions(). Chạy đoạn code dưới đây để xem ví dụ với math.exp().

In [3]:
import math

def basic_sigmoid(x):
    """
    Tính sigmoid của x.

    Đối số:
    x -- Số vô hướng

    Trả về:
    s -- sigmoid(x)
    """
    
    ### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 1 dòng code)
    s = 1/(1+math.exp(-x))
    ### KẾT THÚC CODE Ở ĐÂY ###
    
    return s

In [4]:
basic_sigmoid(3)

0.9525741268224334

**Kỳ vọng đầu ra**: 
<table style = "width:40%">
    <tr>
    <td>** basic_sigmoid(3) **</td> 
        <td>0.9525741268224334 </td> 
    </tr>

</table>

Trên thực tế, chúng ta hiếm khi sử dụng thư viện "math" trong học sâu vì đầu vào của các hàm là số thực. Trong học sâu, chúng ta chủ yếu sử dụng ma trận và vectơ. Đây là lý do tại sao numpy hữu ích hơn.

In [5]:
### Một nguyên do chúng ta sử dụng "numpy" thay vì "math" trong DL ###
x = [1, 2, 3]
basic_sigmoid(x) # các bạn sẽ thấy điều này gây ra lỗi khi chạy nó, vì x là một vectơ.

TypeError: bad operand type for unary -: 'list'

Thực tế, nếu $ x = (x_1, x_2, ..., x_n) $ là một vectơ hàng thì $ np.exp (x) $ sẽ áp dụng hàm mũ cho mọi phần tử của x. Do đó, kết quả đầu ra sẽ là: $ np.exp (x) = (e^{x_1}, e^{x_2}, ..., e^{x_n})$

In [6]:
import numpy as np

# ví dụ về np.exp
x = np.array([1, 2, 3])
print(np.exp(x)) # kết quả là (exp(1), exp(2), exp(3))

[ 2.71828183  7.3890561  20.08553692]


Hơn nữa, nếu x là một vectơ thì phép toán Python như $ s = x + 3 $ hoặc $ s = \frac {1} {x}$ sẽ xuất s dưới dạng một vectơ có cùng kích thước với x.

In [7]:
# ví dụ về phép toán vectơ
x = np.array([1, 2, 3])
print (x + 3)

[4 5 6]


Bất kỳ lúc nào cần thêm thông tin về hàm numpy, hãy xem [tài liệu chính thức](https://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.exp.html). 

Chúng ta cũng có thể tạo một cell mới trong notebook và viết `np.exp?` (chẳng hạn) để truy cập nhanh vào tài liệu.

**Task 3**: Thực thi hàm sigmoid dùng numpy. 

**Hướng dẫn**: x bây giờ có thể là một số thực, một vectơ hoặc một ma trận. Các cấu trúc dữ liệu chúng ta sử dụng trong numpy biểu diễn các dạng này (vectơ, ma trận,...) được gọi là mảng numpy. Chúng ta không cần phải biết thêm vào lúc này.
$$ \text{For } x \in \mathbb{R}^n \text{,     } sigmoid(x) = sigmoid\begin{pmatrix}
    x_1  \\
    x_2  \\
    ...  \\
    x_n  \\
\end{pmatrix} = \begin{pmatrix}
    \frac{1}{1+e^{-x_1}}  \\
    \frac{1}{1+e^{-x_2}}  \\
    ...  \\
    \frac{1}{1+e^{-x_n}}  \\
\end{pmatrix}\tag{1} $$

In [8]:
import numpy as np # điều này có nghĩa là chúng ta có thể truy cập các hàm numpy bằng cách viết np.functions() thay vì numpy.functions()

def sigmoid(x):
    """
    Tính sigmoid của x

    Đối số:
    x -- Mảng numpy hoặc vô hướng có bất kỳ kích thước nào 

    Trả về:
    s -- sigmoid(x)
    """
    
    ### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 1 dòng code)
    s = 1/(1+np.exp(-x))
    ### KẾT THÚC CODE Ở ĐÂY ###
    
    return s

In [9]:
x = np.array([1, 2, 3])
sigmoid(x)

array([0.73105858, 0.88079708, 0.95257413])

**Kỳ vọng đầu ra**: 
<table>
    <tr> 
        <td> **sigmoid([1,2,3])**</td> 
        <td> array([ 0.73105858,  0.88079708,  0.95257413]) </td> 
    </tr>
</table> 


### 1.2 - Sigmoid gradient

Như đã thấy trong bài giảng, chúng ta sẽ cần tính gradient để tối ưu hóa các hàm mất mát bằng cách sử dụng phương pháp truyền ngược. Hãy mã hóa hàm gradient đầu tiên.

**Task 4**: Triển khai hàm sigmoid_grad() để tính gradient của hàm sigmoid liên quan đến đầu vào x của nó. Công thức là:$$sigmoid\_derivative(x) = \sigma'(x) = \sigma(x) (1 - \sigma(x))\tag{2}$$
Chúng ta thường viết code hàm này theo hai bước:
1. Đặt s là sigmoid của x. Bạn sẽ thấy hàm sigmoid(x) của mình hữu ích.
2. Tính $\sigma'(x) = s(1-s)$

In [10]:
def sigmoid_derivative(x):
    """
    Tính gradient (cũng gọi là slope hoặc đạo hàm) của hàm sigmoid với đầu vào x của nó.
    Lưu đầu ra của hàm sigmoid vào các biến rồi sử dụng nó để tính gradient
    
    Đối số:
    x -- Mảng vô hướng hoặc mảng numpy

    Trả về:
    ds -- Gradient đã tính.
    """
    
    ### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 2 dòng code)
    s = 1/(1+np.exp(-x))
    ds = s * (1-s)
    ### KẾT THÚC CODE Ở ĐÂY ###
    
    return ds

In [11]:
x = np.array([1, 2, 3])
print ("sigmoid_derivative(x) = " + str(sigmoid_derivative(x)))

sigmoid_derivative(x) = [0.19661193 0.10499359 0.04517666]


**Kỳ vọng đầu ra**: 


<table>
    <tr> 
        <td> **sigmoid_derivative([1,2,3])**</td> 
        <td> [ 0.19661193  0.10499359  0.04517666] </td> 
    </tr>
</table> 



### 1.3 - Định hình lại mảng ###

Hai hàm numpy phổ biến được sử dụng trong học sâu là [np.shape](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.shape.html) and [np.reshape()](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html). 
- X.shape được sử dụng để lấy shape (kích thước) của ma trận/vectơ X.
- X.reshape(...) được dùng để định hình lại X sang một số kích thước khác.

Ví dụ, trong khoa học máy tính, hình ảnh được biểu diễn bằng mảng 3D có shape $(length, height, depth = 3)$. Tuy nhiên, khi đọc hình ảnh dưới dạng đầu vào của một thuật toán, bạn sẽ chuyển đổi nó thành một vectơ có shape $(length*height*3, 1)$. Nói cách khác, chúng ta "triển khai" hoặc định hình lại mảng 3D thành vectơ 1D.

<img src="images/image2vector_kiank.png" style="width:500px;height:300;">

**Task 5**: Triển khai `image2vector()` nhận đầu vào của shape (length, height, 3) và trả về một vectơ có shape (length\*height * 3,1). Ví dụ: nếu bạn muốn định hình lại mảng v có shape (a, b, c) thành một vectơ có shape (a * b, c), chúng ta sẽ thực hiện:
``` python
v = v.reshape((v.shape[0]*v.shape[1], v.shape[2])) # v.shape[0] = a ; v.shape[1] = b ; v.shape[2] = c
```
- Vui lòng không mã hóa các kích thước của hình ảnh như một hằng số. Thay vào đó, hãy tra cứu số lượng bạn cần với `image.shape[0]`, v.v.

In [12]:
def image2vector(image):
    """
    Đối số:
    image -- mảng numpy có shape (length, height, depth)
    
    Trả về:
    v -- vectơ có shape (length*height*depth, 1)
    """
    
    ### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 1 dòng code)
    v = image.reshape(image.shape[0] * image.shape[1] * image.shape[2], 1)
    ### KẾT THÚC CODE Ở ĐÂY ###
    
    return v

In [13]:
# Đây là mảng 3x3x2, thông thường hình ảnh sẽ là (num_px_x, num_px_y,3) trong đó 3 biểu diễn các giá trị RGB
image = np.array([[[ 0.67826139,  0.29380381],
        [ 0.90714982,  0.52835647],
        [ 0.4215251 ,  0.45017551]],

       [[ 0.92814219,  0.96677647],
        [ 0.85304703,  0.52351845],
        [ 0.19981397,  0.27417313]],

       [[ 0.60659855,  0.00533165],
        [ 0.10820313,  0.49978937],
        [ 0.34144279,  0.94630077]]])

print ("image2vector(image) = " + str(image2vector(image)))

image2vector(image) = [[0.67826139]
 [0.29380381]
 [0.90714982]
 [0.52835647]
 [0.4215251 ]
 [0.45017551]
 [0.92814219]
 [0.96677647]
 [0.85304703]
 [0.52351845]
 [0.19981397]
 [0.27417313]
 [0.60659855]
 [0.00533165]
 [0.10820313]
 [0.49978937]
 [0.34144279]
 [0.94630077]]


**Kỳ vọng đầu ra**: 


<table style="width:100%">
     <tr> 
       <td> **image2vector(image)** =
        [[ 0.67826139] <br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.29380381] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.90714982] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.52835647] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.4215251 &nbsp] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.45017551] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.92814219] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.96677647] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.85304703] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.52351845] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.19981397] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.27417313] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.60659855] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.00533165] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.10820313] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.49978937] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.34144279] <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[ 0.94630077]]</td> 
     </tr>
    
   
</table>

### 1.4 - Chuẩn hóa các hàng

Một kỹ thuật phổ biến khác mà chúng ta sử dụng trong Machine Learning và Deep Learning là chuẩn hóa dữ liệu. Nó thường dẫn đến chất lượng tốt hơn bởi vì gradient descent hội tụ nhanh hơn sau khi chuẩn hóa. Ở đây, theo cách chuẩn hóa, chúng ta muốn thay đổi x thành $\frac {x} {\| x \|}$ (chia mỗi vectơ hàng của x cho chuẩn của nó).

Ví dụ: nếu $$x = 
\begin{bmatrix}
    0 & 3 & 4 \\
    2 & 6 & 4 \\
\end{bmatrix}\tag{3}$$ thì $$\| x\| = np.linalg.norm(x, axis = 1, keepdims = True) = \begin{bmatrix}
    5 \\
    \sqrt{56} \\
\end{bmatrix}\tag{4} $$và        $$ x\_normalized = \frac{x}{\| x\|} = \begin{bmatrix}
    0 & \frac{3}{5} & \frac{4}{5} \\
    \frac{2}{\sqrt{56}} & \frac{6}{\sqrt{56}} & \frac{4}{\sqrt{56}} \\
\end{bmatrix}\tag{5}$$ Lưu ý rằng chúng ta có thể chia các ma trận có kích thước khác nhau và nó vẫn hoạt động tốt: đây được gọi là broadcasting (truyền phát) và chúng ta sẽ tìm hiểu về nó sau.


**Task 6**: Triển khai normalizeRows() để chuẩn hóa các hàng của ma trận. Sau khi áp dụng hàm này cho ma trận đầu vào x, mỗi hàng của x phải là một vectơ có độ dài đơn vị (nghĩa là độ dài 1).

In [14]:
def normalizeRows(x):
    """
    Triển khai một hàm chuẩn hóa từng hàng của ma trận x (có độ dài đơn vị)
    
    Đối số:
    x -- Ma trận numpy có shape (n, m)
    
    Trả về:
    x -- Ma trận numpy đã chuẩn hóa (theo hàng). Chúng ta được phép sửa x.
    """
    
    ### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 2 dòng code)
    # Tính x_norm là chuẩn 2 của x. Sử dụng np.linalg.norm(..., ord = 2, axis = ..., keepdims = True)
    x_norm = np.linalg.norm(x, ord=2, axis=1, keepdims=True)
    # Chia x cho chuẩn của nó.
    x = x/x_norm
    ### KẾT THÚC CODE Ở ĐÂY ###

    return x

In [15]:
x = np.array([
    [0, 3, 4],
    [1, 6, 4]])
print("normalizeRows(x) = " + str(normalizeRows(x)))

normalizeRows(x) = [[0.         0.6        0.8       ]
 [0.13736056 0.82416338 0.54944226]]


**Kỳ vọng đầu ra**: 

<table style="width:60%">

     <tr> 
       <td> **normalizeRows(x)** </td> 
       <td> [[ 0.          0.6         0.8       ]
 [ 0.13736056  0.82416338  0.54944226]]</td> 
     </tr>
    
   
</table>

**Lưu ý**:
Trong normalizeRows(), chúng ta có thể thử in các shape của x_norm và x, sau đó chạy lại đánh giá. Các bạn sẽ nhận ra rằng chúng có các hình dạng khác nhau. Điều này là bình thường khi x_norm lấy chuẩn của mỗi hàng x. Vì vậy, x_norm có cùng số hàng nhưng chỉ có 1 cột. Vậy nó hoạt động như thế nào khi chia x cho x_norm? Đây được gọi là broadcasting và chúng ta sẽ nói về nó ngay bây giờ!

### 1.5 - Broadcasting (Truyền phát) và hàm softmax ####
Một khái niệm rất quan trọng cần hiểu trong numpy là "broadcasting" (truyền phát). Nó rất hữu ích khi thực hiện các phép toán giữa các mảng có hình dạng khác nhau. Để biết chi tiết đầy đủ về broadcasting, bạn có thể đọc [tài liệu broadcasting] chính thức (http://docs.scipy.org/doc/numpy/user/basics.broadcasting.html).

**Task 7**: Thực hiện một hàm softmax bằng cách sử dụng numpy. Chúng ta có thể coi softmax như một hàm chuẩn hóa được sử dụng khi thuật toán cần phân loại hai hoặc nhiều lớp. Chúng ta sẽ tìm hiểu thêm về softmax trong các bài học tiếp theo.

**Hướng dẫn**:
- $ \text{cho } x \in \mathbb{R}^{1\times n} \text{,     } softmax(x) = softmax(\begin{bmatrix}
    x_1  &&
    x_2 &&
    ...  &&
    x_n  
\end{bmatrix}) = \begin{bmatrix}
     \frac{e^{x_1}}{\sum_{j}e^{x_j}}  &&
    \frac{e^{x_2}}{\sum_{j}e^{x_j}}  &&
    ...  &&
    \frac{e^{x_n}}{\sum_{j}e^{x_j}} 
\end{bmatrix} $ 

- $\text{cho ma trận } x \in \mathbb{R}^{m \times n} \text{,  $x_{ij}$ ánh xạ tới phần tử trong hàng thứ $i^{th}$ và cột thứ $j^{th}$ của $x$, ta được: }$  $$softmax(x) = softmax\begin{bmatrix}
    x_{11} & x_{12} & x_{13} & \dots  & x_{1n} \\
    x_{21} & x_{22} & x_{23} & \dots  & x_{2n} \\
    \vdots & \vdots & \vdots & \ddots & \vdots \\
    x_{m1} & x_{m2} & x_{m3} & \dots  & x_{mn}
\end{bmatrix} = \begin{bmatrix}
    \frac{e^{x_{11}}}{\sum_{j}e^{x_{1j}}} & \frac{e^{x_{12}}}{\sum_{j}e^{x_{1j}}} & \frac{e^{x_{13}}}{\sum_{j}e^{x_{1j}}} & \dots  & \frac{e^{x_{1n}}}{\sum_{j}e^{x_{1j}}} \\
    \frac{e^{x_{21}}}{\sum_{j}e^{x_{2j}}} & \frac{e^{x_{22}}}{\sum_{j}e^{x_{2j}}} & \frac{e^{x_{23}}}{\sum_{j}e^{x_{2j}}} & \dots  & \frac{e^{x_{2n}}}{\sum_{j}e^{x_{2j}}} \\
    \vdots & \vdots & \vdots & \ddots & \vdots \\
    \frac{e^{x_{m1}}}{\sum_{j}e^{x_{mj}}} & \frac{e^{x_{m2}}}{\sum_{j}e^{x_{mj}}} & \frac{e^{x_{m3}}}{\sum_{j}e^{x_{mj}}} & \dots  & \frac{e^{x_{mn}}}{\sum_{j}e^{x_{mj}}}
\end{bmatrix} = \begin{pmatrix}
    softmax\text{(hàng thứ nhất của x)}  \\
    softmax\text{(hàng thứ hai của x)} \\
    ...  \\
    softmax\text{(hàng cuối của x)} \\
\end{pmatrix} $$

In [16]:
def softmax(x):
    """Tính softmax cho từng hàng của đầu vào x.

    Code của chúng ta phải hoạt động cho vectơ hàng và cả cho ma trận shape (n, m).

    Đối số:
    x -- Ma trận numpy có shape (n,m)

    Trả về:
    s -- Ma trận numpy bằng với softmax của x, của shape (n,m)
    """
    
    ### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 3 dòng code)
    # Áp dụng element-wise exp() vào x. Sử dụng np.exp(...).
    x_exp = np.exp(x)


    # Tạo vetơ x_sum tính tổng từng hàng của x_exp. Sử dụng np.sum(..., axis = 1, keepdims = True).
    x_sum = np.sum(x_exp, axis=1, keepdims=True)
    # Tính softmax(x) bằng cách chia  x_exp cho x_sum. Nó sẽ được tự động hóa bằng numpy broadcasting.
    s = x_exp / x_sum
    ### KẾT THÚC CODE Ở ĐÂY ###
    
    return s

In [17]:
x = np.array([
    [9, 2, 5, 0, 0],
    [7, 5, 0, 0 ,0]])
print("softmax(x) = " + str(softmax(x)))

softmax(x) = [[9.80897665e-01 8.94462891e-04 1.79657674e-02 1.21052389e-04
  1.21052389e-04]
 [8.78679856e-01 1.18916387e-01 8.01252314e-04 8.01252314e-04
  8.01252314e-04]]


**Kỳ vọng đầu ra**:

<table style="width:60%">

     <tr> 
       <td> **softmax(x)** </td> 
       <td> [[  9.80897665e-01   8.94462891e-04   1.79657674e-02   1.21052389e-04
    1.21052389e-04]
 [  8.78679856e-01   1.18916387e-01   8.01252314e-04   8.01252314e-04
    8.01252314e-04]]</td> 
     </tr>
</table>


**Lưu ý**:
- Nếu in shape của x_exp, x_sum và s ở trên và chạy lại cell đánh giá, chúng ta sẽ thấy rằng x_sum có shape (2,1) trong khi x_exp và s có shape (2,5). **x_exp/x_sum** hoạt động nhờ Python broadcasting.

Xin chúc mừng! Bây giờ các bạn đã hiểu khá tốt về numpy Python và đã triển khai một số hàm hữu ích mà chúng ta sẽ sử dụng trong Deep learning.

<font color='blue'>

**Những điều cần nhớ:**
- np.exp(x) hoạt động với bất kỳ np.array x nào và áp dụng hàm mũ cho mọi tọa độ
- hàm sigmoid và gradient của nó
- image2vector thường được sử dụng trong deep learning
- np.reshape được sử dụng rộng rãi. Trong tương lai, các bạn sẽ thấy rằng việc giữ thẳng kích thước ma trận/vectơ sẽ giúp loại bỏ rất nhiều lỗi
- numpy có các hàm tích hợp hiệu quả
- broadcasting cực kỳ hữu ích

## 2) Vectơ hóa


Trong deep learning, chúng ta phải xử lý các tập dữ liệu rất lớn. Do đó, một hàm không tối ưu về mặt tính toán có thể trở thành một nút thắt lớn trong thuật toán và có thể khiến mô hình mất nhiều thời gian để chạy. Để đảm bảo code hiệu quả về mặt tính toán, hãy sử dụng vectơ hóa. Ví dụ: cố gắng phân biệt sự khác biệt giữa các cách triển khai sau đây của tích vô hướng/tích ngoài/elementwise.

In [18]:
import time

x1 = [9, 2, 5, 0, 0, 7, 5, 0, 0, 0, 9, 2, 5, 0, 0]
x2 = [9, 2, 2, 9, 0, 9, 2, 5, 0, 0, 9, 2, 5, 0, 0]

### TÍCH VÔ HƯỚNG CỦA TRIỂN KHAI VECTƠ ###
tic = time.process_time()
dot = 0
for i in range(len(x1)):
    dot+= x1[i]*x2[i]
toc = time.process_time()
print ("dot = " + str(dot) + "\n ----- Computation time = " + str(1000*(toc - tic)) + "ms")

### TÍCH NGOÀI CỦA TRIỂN KHAI VECTƠ ###
tic = time.process_time()
outer = np.zeros((len(x1),len(x2))) # we create a len(x1)*len(x2) matrix with only zeros
for i in range(len(x1)):
    for j in range(len(x2)):
        outer[i,j] = x1[i]*x2[j]
toc = time.process_time()
print ("outer = " + str(outer) + "\n ----- Computation time = " + str(1000*(toc - tic)) + "ms")

### TRIỂN KHAI ELEMENTWISE ###
tic = time.process_time()
mul = np.zeros(len(x1))
for i in range(len(x1)):
    mul[i] = x1[i]*x2[i]
toc = time.process_time()
print ("elementwise multiplication = " + str(mul) + "\n ----- Computation time = " + str(1000*(toc - tic)) + "ms")

### TRIỂN KHAI TÍCH VÔ HƯỚNG TỔNG QUÁT ###
W = np.random.rand(3,len(x1)) # Mảng numpy 3*len(x1) ngẫu nhiên
tic = time.process_time()
gdot = np.zeros(W.shape[0])
for i in range(W.shape[0]):
    for j in range(len(x1)):
        gdot[i] += W[i,j]*x1[j]
toc = time.process_time()
print ("gdot = " + str(gdot) + "\n ----- Computation time = " + str(1000*(toc - tic)) + "ms")

dot = 278
 ----- Computation time = 0.11231200000005881ms
outer = [[81. 18. 18. 81.  0. 81. 18. 45.  0.  0. 81. 18. 45.  0.  0.]
 [18.  4.  4. 18.  0. 18.  4. 10.  0.  0. 18.  4. 10.  0.  0.]
 [45. 10. 10. 45.  0. 45. 10. 25.  0.  0. 45. 10. 25.  0.  0.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [63. 14. 14. 63.  0. 63. 14. 35.  0.  0. 63. 14. 35.  0.  0.]
 [45. 10. 10. 45.  0. 45. 10. 25.  0.  0. 45. 10. 25.  0.  0.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [81. 18. 18. 81.  0. 81. 18. 45.  0.  0. 81. 18. 45.  0.  0.]
 [18.  4.  4. 18.  0. 18.  4. 10.  0.  0. 18.  4. 10.  0.  0.]
 [45. 10. 10. 45.  0. 45. 10. 25.  0.  0. 45. 10. 25.  0.  0.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0

In [19]:
x1 = [9, 2, 5, 0, 0, 7, 5, 0, 0, 0, 9, 2, 5, 0, 0]
x2 = [9, 2, 2, 9, 0, 9, 2, 5, 0, 0, 9, 2, 5, 0, 0]

### TÍCH VÔ HƯỚNG VECTƠ HÓA CỦA VECTƠ ###
tic = time.process_time()
dot = np.dot(x1,x2)
toc = time.process_time()
print ("dot = " + str(dot) + "\n ----- Computation time = " + str(1000*(toc - tic)) + "ms")

### TÍCH NGOÀI VECTƠ HÓA ###
tic = time.process_time()
outer = np.outer(x1,x2)
toc = time.process_time()
print ("outer = " + str(outer) + "\n ----- Computation time = " + str(1000*(toc - tic)) + "ms")

### PHÉP NHÂN ELEMENTWISE VECTƠ HÓA  ###
tic = time.process_time()
mul = np.multiply(x1,x2)
toc = time.process_time()
print ("elementwise multiplication = " + str(mul) + "\n ----- Computation time = " + str(1000*(toc - tic)) + "ms")

### TÍCH VÔ HƯỚNG TỔNG QUÁT VECTƠ HÓA ###
tic = time.process_time()
dot = np.dot(W,x1)
toc = time.process_time()
print ("gdot = " + str(dot) + "\n ----- Computation time = " + str(1000*(toc - tic)) + "ms")

dot = 278
 ----- Computation time = 0.06752099999995931ms
outer = [[81 18 18 81  0 81 18 45  0  0 81 18 45  0  0]
 [18  4  4 18  0 18  4 10  0  0 18  4 10  0  0]
 [45 10 10 45  0 45 10 25  0  0 45 10 25  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [63 14 14 63  0 63 14 35  0  0 63 14 35  0  0]
 [45 10 10 45  0 45 10 25  0  0 45 10 25  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [81 18 18 81  0 81 18 45  0  0 81 18 45  0  0]
 [18  4  4 18  0 18  4 10  0  0 18  4 10  0  0]
 [45 10 10 45  0 45 10 25  0  0 45 10 25  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]]
 ----- Computation time = 0.0541269999999594ms
elementwise multiplication = [81  4 10  0  0 63 10  0  0  0 81  4 25  0  0]
 ----- Computation time = 0.03317700000005086ms
gdot = [24.78942909 23.03162455 26.5536801

Như các bạn thấy, việc triển khai vectơ hóa nhanh và hiệu quả hơn nhiều. Đối với các vectơ/ma trận lớn hơn, sự khác biệt về thời gian chạy thậm chí còn lớn hơn.

**Lưu ý** rằng `np.dot()` thực hiện phép nhân ma trận-ma trận hoặc ma trận-vectơ. Điều này khác với toán tử `np.multiply()` và `*` (tương đương với `.*` trong Matlab/Octave), thực hiện phép nhân từng phần tử của 2 ma trận cùng chiều.

### 2.1 Thực hiện các hàm mất mát L1 và L2

**Task 8**: Thực hiện phiên bản vectơ hóa numpy của hàm mất mát L1. Chúng ta có thể thấy hàm abs(x) (giá trị tuyệt đối của x) hữu ích.

**Nhắc nhở**:
- Hàm mất mát được sử dụng để đánh giá chất lượng mô hình. Giá trị mất mát càng lớn thì các dự đoán ($\hat {y} $) càng khác với giá trị thực ($y$). Trong deep learning, chúng ta sử dụng các thuật toán tối ưu hóa như Gradient Descent để huấn luyện mô hình và để giảm thiểu chi phí.
- Giá trị mất mát L1 được xác định như sau:
$$\begin{align*} & L_1(\hat{y}, y) = \sum_{i=0}^m|y^{(i)} - \hat{y}^{(i)}| \end{align*}\tag{6}$$

In [20]:
def L1(yhat, y):
    """
    Đối số:
    yhat -- vectơ có kích thước (nhãn dự đoán)
    y -- vectơ có kích thước (nhãn đúng)
    
    Trả về:
    loss -- giá trị của hàm mất mát L1 đã xác định ở trên 
    """
    
    ### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 1 dòng code)
    loss = np.sum(np.abs(y - yhat))
    ### KẾT THÚC CODE Ở ĐÂY ###
    
    return loss

In [21]:
yhat = np.array([.9, 0.2, 0.1, .4, .9])
y = np.array([1, 0, 0, 1, 1])
print("L1 = " + str(L1(yhat,y)))

L1 = 1.1


**Kỳ vọng đầu ra**:

<table style="width:20%">

     <tr> 
       <td> **L1** </td> 
       <td> 1.1 </td> 
     </tr>
</table>


**Task 9**: Thực hiện phiên bản vectơ hóa numpy của giá trị mất mát L2. Có một số cách để thực hiện L2 nhưng các bạn sẽ thấy hàm np.dot() hữu ích. Xin nhắc lại, nếu $x = [x_1, x_2, ..., x_n]$, thì `np.dot(x,x)` = $\sum_{j=0}^n x_j^{2}$. 

- L2 được xác định như sau $$\begin{align*} & L_2(\hat{y},y) = \sum_{i=0}^m(y^{(i)} - \hat{y}^{(i)})^2 \end{align*}\tag{7}$$

In [22]:
def L2(yhat, y):
    """
    Đối số:
    yhat -- vectơ có kích thước (nhãn dự đoán)
    y -- vectơ có kích thước (nhãn đúng)
    
    Trả về:
    loss -- giá trị của hàm mất mát L2 đã xác định ở trên
    """
    
    ### BẮT ĐẦU CODE Ở ĐÂY ### (≈ 1 dòng code)
    loss = np.sum(np.power(y-yhat, 2))
    ### KẾT THÚC CODE Ở ĐÂY ###
    
    return loss

In [23]:
yhat = np.array([.9, 0.2, 0.1, .4, .9])
y = np.array([1, 0, 0, 1, 1])
print("L2 = " + str(L2(yhat,y)))

L2 = 0.43


**Kỳ vọng đầu ra**: 
<table style="width:20%">
     <tr> 
       <td> **L2** </td> 
       <td> 0.43 </td> 
     </tr>
</table>

Chúc mừng các bạn đã hoàn thành bài tập này. Chúng tôi hy vọng rằng bài tập khởi động nhỏ này sẽ giúp ích cho các bạn trong những bài tập sắp tới, trở nên hào hứng và thú vị hơn!

<font color='blue'>

**Những điều cần nhớ:**
- Vectơ hóa là rất quan trọng trong deep learning. Nó giúp tính toán chất lượng và rõ ràng hơn.
- Chúng ta đã xem lại hàm mất mát L1 và L2.
- Chúng ta đã quen với nhiều hàm numpy như np.sum, np.dot, np.multiply, np.maximum,...